keps/sig-api-machinery/storage-migration-auto-trigger.md

```diff
@@ -225,6 +208,137 @@ manually, e.g., deploying the [initializer][] manually.
+| 10         | Triggering controllers update currentSV. The leader does not start migration because of the lack of consensus on storage version.  | v2                        | v1             | v1             | v2, v1, v1            | v1, v2                  |
+| 12         | Rolling upgrade continues, apiserver 2 is upgraded.                                                                                | v2                        | v2             | v1             | v2, v1, v1            | v1, v2                  |
+| 18         | Rolling upgrade is done, apiserer 3 is upgraded.                                                                                   | v2                        | v2             | v2             | v2, v1, v1            | v1, v2                  |
+| 20         | Triggering controllers update currentSV. The leader starts migration as apiservers have the same storage version.                  | v2                        | v2             | v2             | v2, v2, v2            | v1, v2                  |
+| 22         | The cluster admin rolls back apiserver 3.                                                                                          | v2                        | v2             | v1             | v2, v2, v2            | v1, v2                  |
+| 30         | The triggering controller co-locates with apiserver 3 adds v1 to currentSV and persistedSV. The leader cancels ongoing migration.  | v2                        | v2             | v1             | v2, v2, v1            | v1, v2                  |
+| 34         | The cluster admin upgrades apiserver 3 again.                                                                                      | v2                        | v2             | v2             | v2, v2, v1            | v1, v2                  |
+| 40         | The triggering controller co-locates with apiserver 3 sets currentSV to v2. The leader starts migration as consensus is reached.   | v2                        | v2             | v2             | v2, v2, v2            | v1, v2                  |
+| 45         | The migration has completed. The leader updates persistedSV to v2.                                                                 | v2                        | v2             | v2             | v2, v2, v2            | v2                      |
+
+1. sv: storage version.
+2. CurrentSV: currentStorageVersionHashes.
+3. PersistedSV: persistedStorageVersionHashes.
+
+### Pros and Cons
+Pros:
+1. The API changes are limited to the storage migration API, which is only used
+   by the migrator, so we have more flexibility to experiment.
+
+Cons:
+1. There are two assumptions in this design that might not hold in the future
+    * there is a triggering controller co-locates with every apiserver that
+      consists the HA master.
+    * the triggering controller only talks with the local apiserver.
```

Comment:
If we ever fix apiserver to coordinate internally (e.g. publish a constant discovery doc), will that break this storage migrator design?

Reply:
Not sure what you meant by "constant discovery doc".

I think it depends on how the apiservers are fixed.
* If HA apiservers can flip storage versions at the same time, then no change is required to the storage migrator.
* If HA apiservers show a consistent discovery doc, and marks the storage version as "undefined" when there is no consensus, then the storage migrator needs to be updated to handle the "undefined" state.

In both cases, this design doesn't add additional difficulty.